domain/credential/state/credential.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/cloud"
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/database"
+	"github.com/juju/juju/domain"
+	dbcloud "github.com/juju/juju/domain/cloud/state"
+)
+
+// State is used to access the database.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState creates a state to access the database.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// UpsertCloudCredential adds or updates a cloud credential with the given tag.
+func (st *State) UpsertCloudCredential(ctx context.Context, name, cloud, owner string, credential cloud.Credential) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get the credential UUID - either existing or make a new one.
+		// TODO(wallyworld) - implement owner as a FK when users are modelled.
+
+		q := `
+SELECT  cloud_credential.uuid AS &M.uuid
+FROM    cloud_credential
+        JOIN cloud ON cloud_credential.cloud_uuid = cloud.uuid
+WHERE   cloud.name = $M.cloud_name AND cloud_credential.name = $M.credential_name AND cloud_credential.owner_uuid = $M.owner
+`
+		stmt, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		result := sqlair.M{}
```

Comment:
I really wish something other than just map or struct were supported to select a result into. To get a scalar string for example, you need to use a map and then get the key from the map etc. Compare this with the raw SQL case of simply scanning into a string var.

Reply:
Update: this is on the radar to be addressed